src/vcpkg/base/system.mac.cpp

```diff
@@ -0,0 +1,251 @@
+#include <vcpkg/base/hash.h>
+#include <vcpkg/base/parse.h>
+#include <vcpkg/base/system.mac.h>
+#include <vcpkg/base/system.process.h>
+#include <vcpkg/base/util.h>
+
+#include <map>
+
+#if !defined(_WIN32)
+#include <net/if.h>
+#if defined(__APPLE__)
+#include <ifaddrs.h>
+
+#include <net/if_dl.h>
+#define AF_TYPE AF_LINK
+#elif defined(__linux__)
+#include <ifaddrs.h>
+
+#include <netpacket/packet.h>
+#define AF_TYPE AF_PACKET
+#else
+// Fallback to use ioctl calls for systems without getifaddrs()
+#include <unistd.h>
+
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#endif
+#endif
+
+namespace vcpkg
+{
+    static constexpr size_t MAC_BYTES_LENGTH = 6;
+    // 6 hex-bytes (12 hex-digits) + five separators
+    static constexpr size_t MAC_STRING_LENGTH = MAC_BYTES_LENGTH * 2 + 5;
+
+    bool validate_mac_address_format(StringView mac)
+    {
+        static constexpr char format[] = "xx:xx:xx:xx:xx:xx";
+        if (mac.size() != MAC_STRING_LENGTH) return false;
+        for (size_t i = 0; i < MAC_STRING_LENGTH; ++i)
+        {
+            if (format[i] == ':' && mac[i] != ':') return false;
+            if (format[i] == 'x' && !ParserBase::is_hex_digit(mac[i])) return false;
+        }
+        return true;
+    }
+
+    bool is_valid_mac_for_telemetry(StringView mac)
+    {
+        // this exclusion list is the taken from VS Code's source code
+        // https://github.com/microsoft/vscode/blob/main/src/vs/base/node/macAddress.ts
+        static constexpr std::array<StringLiteral, 3> invalid_macs{
+            "00:00:00:00:00:00",
+            "ff:ff:ff:ff:ff:ff",
+            // iBridge MAC address used on some Apple devices
+            "ac:de:48:00:11:22",
+        };
+
+        if (!validate_mac_address_format(mac)) return false;
+        return !Util::Vectors::contains(invalid_macs, mac);
+    }
+
+    std::string mac_bytes_to_string(const Span<unsigned char>& bytes)
+    {
+        if (bytes.size() != MAC_BYTES_LENGTH) return "";
+
+        static constexpr char hexits[] = "0123456789abcdef";
+        char mac_address[MAC_STRING_LENGTH];
+        char* mac = mac_address;
+        unsigned char c = static_cast<unsigned char>(bytes[0]);
```

Comment:
These static_casts do nothing now that the input is `Span<unsigned char>` (but the right thing to do may be to accept char rather than unsigned char here)